### ISSUE TITLE
implement NDELAY option

### ISSUE THREAD

**Issue by goatshriek** [goatshriek/stumpless#79]
Stumpless currently does not support the STUMPLESS_OPTION_NDELAY option despite providing it in its header files. This option is meant to mirror the LOG_NDELAY option in functionality. The meaning may change across target types though - this will need to be thought through, defined, and documented for each of the target types.

Don't forget to document the option after you implement it in `include/stumpless/option.h`.

Check out the [Contributing Guidelines](https://github.com/goatshriek/stumpless/blob/latest/docs/CONTRIBUTING.md) and the [development guide](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md) for the basics on working with stumpless and submitting changes.

**Comment by iamsb97** [comment]
@goatshriek : If this is still within the scope, I would like to give this a try.

**Comment by goatshriek** [comment]
It is still open for work, please feel free to work through it!

For this issue in particular, it is fine to start by supporting one or a few targets instead of all of them. Don't feel like you need to figure it out for every target type before opening a PR.

**Comment by iamsb97** [comment]
I was looking at LOG_NDELAY behavior and the code for various targets. This is what I understand so far.
NDELAY basically establishes connection to the target when the target is opened and it's counterpart ODELAY does it lazily when writing the first log message.

The current behavior of various targets and what might be required potentially:
- Buffer target - Buffer generated and passed by the user. NDELAY option not applicable.
- File target - Logging file is opened in append mode when stumpless_open_file_target is called. NDELAY is already the default.
- Function target - Function generated and passed by the user. Logging specifics inside the function. NDELAY option not applicable.
- Journald target - Establishes connection at every log message inside sd_journal_sendv. NDELAY option not applicable.
    - Currently, static field initializations (init_fields/init_fixed_fields) are done lazily when sendto_<> is called. That can be done when journald target is created if NDELAY option is specified. 
- Socket target - Unix socket bind done while creating a socket target. NDELAY is already the default.
- Stream target - Stream created and passed by the user. NDELAY option not applicable.
- WEL target - RegisterEventSource is the equivalent of establishing connection which is done when target is created. NDELAY is already the default.
- SQLite3 target - Database connection is established by the user or when stumpless_open_sqlite3_target is called. NDELAY is already the default.
- Chain target - Each target in the chain will have it's behavior defined. NDELAY option not applicable.
- Network target - Network connection established when stumpless_open_network_target is called. NDELAY is already the default.

In short, wherever applicable, NDELAY already seems to be the default behavior. If the above definitions for each target type looks good to you, then a documentation change would be sufficient for this. Please let me know what you think about it.

**Comment by goatshriek** [comment]
Yes, all of these are correct! Thank you for going through the code carefully to confirm this.

Your point about the journald targets is correct as far as it could be done earlier, and in fact I think it might even be possible to do with initialization of a static (thread local) structure and avoid the heap allocation and manual initialization entirely, regardless of the option. I'll create a separate issue to explore the implementation of that improvement somewhere down the road. There's already a valgrind exception for this code, which is something I'd like to remove anyway.

One note on the network targets - currently there are a set of `new` constructors that open targets without initiating a network connection, which would match ODELAY behavior, while the `open` constructors match NDELAY. Combining them into a single constructor and using the option to differentiate would probably be too messy as the `new` constructors don't require all of the parameters an immediate open would, and would preclude users from using a builder-style pattern if that's what they prefer. I think it's worth noting in the option documentation that this paradigm exists for network targets.

I originally intended to use that same pattern for other targets as they began to support the two modes. I welcome your thoughts on that, of course - perhaps there is a better way to go about it?

You're spot on that a documentation update should resolve the NDELAY option issue here. Some code changes may be needed to make sure this option is set on new targets, with the exception of network targets from the `new` family as mentioned above.

**Comment by iamsb97** [comment]
`new` constructors for network target don't really match ODELAY behavior, right? Even when using new constructors, the user would require to use `open` constructor himself later before logging any message whereas ODELAY would initiate network connection when the first message is logged.


On an unrelated note, I see there are 2 constructors for WEL targets (remote and local) with completely identical code except the arguments to `new_wel_target` function. Can these two functions be clubbed into 1 with two arguments to the common function? If you are okay with this, I will open a new issue to handle this.

**Comment by goatshriek** [comment]
Ah you're correct, there is an extra call to `stumpless_open_target` that is needed, which doesn't match ODELAY semantics. My apologies for not catching that.

I've kept the two WEL target constructors separate so that users only needing to interact with a local machine don't need to understand and pass a second argument that's only needed for the remote case. However, there is definitely an opportunity to refactor the local constructor to simply call the remote one with NULL behind the scenes, instead of existing as an entirely separate set of code. Please feel free to open a new issue and/or PR to refactor the implementation of `stumpless_open_local_wel_target` if you'd like - it certainly would be much appreciated!

**Comment by iamsb97** [comment]
Great, thanks! I will open a new issue for refactoring local WEL constructor. 

To summarize, as a fix to this issue, I will update the documentation and describe the behavior for each target and leave all the code implementations as a part of #80.

**Comment by goatshriek** [comment]
Addressed by #490.
